add new counts columns and modified routes queries

NakulManchanda opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

https://github.com/hicommonwealth/commonwealth/wiki/Pre-Aggregated-Metrics

## Link to Issue
Closes: https://github.com/hicommonwealth/commonwealth/issues/4201

## Description of Changes
- New columns migration script
- Compute current count to backfill values
- Modified queries based on new column
-  Add hooks to keep counts upto date


### Support for Scheduled job:
```
yarn --cwd packages/commonwealth recompute-counts
```

## Impact
- Speed up route queries - viewUserActivity, globalUserActivity, getThreads, bulkThreads, activeThreads

## Test Plan
- CA (click around) tested on local and frack:
- Load testing to measure performance impact

## Steps for Testing
- login
- go to any community, select thread
   - add/remove reaction to new or existing Thread(s)
   - add/remove new comment to Thread(s)
   - add/remove reaction to comment(s)
   - check all views(topic view, overview) wherever we show comment & reaction count if they are all updated
   - refresh window again - to verify changes are not done on client side only, and are successfully done on backend
- go to user dashboard
   - verify global & user activity feed to show new thread & new comment activity
   - verify new activity in feed.

## Deployment Plan
1. Heroku scheduler new job - to recompute 
- Click add job - choose 03:00AM <timezone> everyday
- Add this job
```
# instructions to run from root folder
yarn --cwd packages/commonwealth recompute-counts
```

## Other Considerations
- we verified job on heroku scheduler on frick - this statement will work
```
yarn --cwd packages/commonwealth recompute-counts
```

NakulManchanda commented:
Will merge this first - https://github.com/hicommonwealth/commonwealth/pull/4179

NakulManchanda commented:
Time for migrations to run on Frack

== 20230712203803-add-counts-thread-comment-reaction: migrated (44.982s)
== 20230713140523-track-count-triggers: migrated (0.054s)

NakulManchanda commented:
[performance-comparision-master-vs-track-counts.zip](https://github.com/hicommonwealth/commonwealth/files/12045846/performance-comparision-master-vs-track-counts.zip)


**Summary:**

Master:
<img width="619" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/abe6b996-496c-48f5-b619-df967f8baa3c">

NakulManchanda commented on packages/commonwealth/server/util/activityUtils.ts:
```diff
@@ -0,0 +1,84 @@
+         nts.category_id,
+         thr.view_count as view_count,
+         thr.comment_count AS comment_count
+  FROM ranked_thread_notifs nt
```
removed reaction_count - we are not using that in UI

NakulManchanda commented:
After more viewUserActivity changes

`viewUserActivity` 
- min to 77ms, max to 278ms 
- before latest change it was min 163 max 486
- master min 1.1sec 2.7sec


<img width="624" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/4edbf2fa-c6fd-4ba4-a39f-68fdc6c35967">

NakulManchanda commented:
`viewUserActivity`

**Earlier**
```
{
    "status": "Success",
    "result": [
        {
            "thread_id": 12197,
            "last_activity": "2023-07-16T20:44:46.818Z",
            "notification_data": "{\"created_at\":\"2023-07-16T20:44:45.950Z\",\"thread_id\":12197,\"root_type\":\"discussion\",\"root_title\":\"https%3A%2F%2Fcommonwealth.im%2Fethereum%2Fnew%2Fdiscussion\",\"comment_text\":\"%5Bwc%3A43763c828aec6f88f3af170f3700b2a24a033df6260eb4f9ae0c658d73123a97%402%3Frelay-protocol%3Dirn%26symKey%3D8993df9d0262ab715e09dbbebe6066e72b31d156f479f1c196d8f5669668a1d3%5D(https%3A%2F%2F)%0A%0A\",\"chain_id\":\"ethereum\",\"author_address\":\"0xC95e7CD8AAc621f5fE262095febc7170BE4D58A8\",\"author_chain\":\"ethereum\"}",
            "category_id": "new-thread-creation",
            "view_count": 6,
            "comment_count": "0",
            "reaction_count": "0",
            "commenters": []
        },
        {
            "thread_id": 12185,
            "last_activity": "2023-07-15T14:13:00.869Z",
            "notification_data": "{\"created_at\":\"2023-07-15T14:12:59.518Z\",\"thread_id\":12185,\"root_title\":\"%3Ctr%3E%3Ctd%3E%3Cdiv%20align%3D%22center%22%3E%3Ca%20href%3D%22javascript%3Aconsole.log(document.cookie)%22%3EHackerOne%20%3C%7Btag%7D%7Bfiller%7D%7Bevent_handler%7D%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Bjavascript%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%20%3CdEv%20-%20If%20fails%2C%20%3C%5Ba-zA-Z%5D%2B%20confirm(1)%20%3Cd3V%20-%20If%20fails%2C%20%3C%5Ba-zA-Z0-9%5D%2Balert(1)%20%20%3Cd%7C3v%20-%20If%20fails%2C%20%3C.%2Balert(1)%20%3Csvg%20-%20If%20passes%2C%20no%20tag%20checking%20is%20in%20place%20%3Cdev%20-%20If%20fails%2C%3C%5Ba-z%5D%2Balert(1)%20x%3Cdev%20-%20If%20passes%2C%5E%3C%5Ba-z%5D%2Bconfirm(1)%20%3Ctag%250dxxx%3E%20-%20If%20fails%2C%20%5B%5Cs%5Cn%5Cr%2B%5D%2Bconfirm(1)%20%3Ctag%2F~%2Fxxx%20-%20If%20fails%2C%20.%2B%20%3Ctag%20xxx%20-%20If%20fails%2C%20%7Bspace%7D%20%3Ctag%2509xxx%20-%20if%20fails%2C%20%5B%5Cs%5D%20%3Ctag%2509%2509xxx%20-%20if%20fails%2C%20%5Cs%2Balert(1)%20%3Ctag%2Fxxx%20-%20If%20fails%2C%20%5B%5Cs%2F%5D%2Bconfirm(1)%20on%5Cw%2B%3Dconfirm(1)%20on(load%7Cclick%7Cerror%7Cshow)%3Dconfirm(1)%20%3Ctag%7Bfiller%7Donxxx%20-%20If%20fails%2C%20on%5Cw%2B.%20If%20passes%2C%20on(load%7Cclick%7Cerror%7Cshow)%20%3Ctag%7Bfiller%7Donclick%20-%20If%20passes%2C%20no%20event%20handler%20checking%20regular%20expression%20is%20in%20place%20%3Cpayload%250aconfirm(1)%20%3Cpayload%250d%20prompt(1)%20%3Cpayload%2509alert(1)%20%3Cpayload%3Eprompt(1)%20%3Cpayload%20alert(1)%20%3Cpayload%7Bspace%7Dalert(1)%20%3Cpayload%2F%2Falert(1)%20%3Ctag%7Bprompt(1)%7D%7Bconfirm(2)%7D%3E%20%3Ca%20href%3D'https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io'%20onload%3Dconfirm(1)%3E%20%3CsCriPt%7Bfiller%7DsRc%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Burl%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%20%3CobJecT%7Bfiller%7Ddata%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Burl%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%20href%5B%5Cs%5D%7B0%2C%7D%3D%5B%5Cs%5D%7B0%2C%7Djavascript%3Aonoload%3Dconfirm(1)%20%3CA%7Bfiller%7DhReF%7B%3Ffiller%7D%3D%7B%3Ffiller%7DJavaScript%3A%7Bjavascript%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%20%3CA%7Bfiller%7DhReF%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Bquote%7D%7Bspecial%7D%3A%7Bjavascript%7D%7Bquote%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%20j%250aAv%250dasCr%2509ipt%3Aconfirm(1)%20J%250aa%250av%250aa%250as%250ac%250ar%250ai%250ap%250aT%250a%3Aprompt(1)%20J%250aa%250dv%2509a%250as%250dc%2509r%250ai%250dp%2509T%250d%250a%3Aalert(1)%20%26%2374%3Bavascript%26colon%3Bconfirm(1)%20jav%26%23x61%3B%26%23115%3Bcript%3Aconfirm(1)%20%26%2374%3Bava%250a%250d%2509script%26colon%3B%20%20%3C--%24input--%3Ealert(1)%3C--%2F%24input--%3E%20%3C%2Ftag%3Ealert(1)%20%3C%2FtAg%2Fx%3Eprompt(1)%20%3C%2Ftag%7Bspace%7D%3Eprompt(1)%20%3C%2Ftag%2F%2F%3Econfirm(1)%20%3C%2Ftag%250a%3Econfirm(1)%20%3C%2Ftag%250d%3Ealert(1)%20%3C%2Ftag%2509%3Eonerror%3Dalert(1)%20%7Bclosing%20tag%7D%7Balert(1)%7D%20%3Cinput%20value%3D%22%24input%22%20type%3D%22text%22%3E%20%3Cscript%20src%3D%22%24input%22%3E%20%3Cscript%20src%3D%22https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%22%3E%20%3Ciframe%20srcdoc%3D%22%24input%22%3E%20%2F%2Fhttps%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%20bypasses%20http(%3Fs)%3A%2F%2F%20%2F%2F%2F%2F%2F%2F%2F%2Fhttps%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%20bypasses%20(%3F%3Ahttp(%3Fs)%3A%3F)%3F%2F%2F%20%2F%5C%2F%2F%2F%5C%5C%2Fhttps%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%20bypasses%20(%3F%3Ahttp(%3Fs)%3A%3F)%3F%2F%2F%2B%20%3Ciframe%20srcdoc%3D%22%26lt%3Bsvg%2Fonload%3Dalert()%26gt%3B%22%3E%20%3Cinput%20type%3D%22text%22%20value%3D%22%22%2Fonfocus%3D%22alert()%24input%22%3E%20%7Bquote%7D%7Bfiller%7D%7Bevent_handler%7D%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Bjavascript%7D%20%7Bquote%7D%3E%7Bonclick%3Dconfirm(1)%7D%20var%20name%20%3D%20'%24input'%3B%20%7Bquote%7D%7Bdelimiter%7D%7Bjavascript%7D%7Bdelimiter%7D%7Bquote%7D%20'%25%7Bjavascript%7D%25'confirm(1)'%25%7B%2Fjavascript%7D%25'%20'%7C%7Bjavascript%7D%7C'alert(1)'%7C%7B%2Fjavascript%7D%7C'%20'%3C%7Bjavascript%7D%3C'prompt(1)'%3C%7B%2Fjavascript%7D%3C%20'%3E%7Bjavascript%7D%3E'confirm(1)'%3C%2F%7Bjavascript%7D%3C%20'%5E%7Bjavascript%7D%5E'alert(1)'%3C%2F%7Bjavascript%7D%3C%20'*%7Bjavascript%7D*'prompt(1)'%3C%7B%2Fjavascript%7D%3C%20'%2B%7Bjavascript%7D%2B'confirm(1)'%2B%2F%7Bjavascript%7D%2B'%20'%2F%7Bjavascript%7D%2F'alert(1)'%2F%7Bjavascript%7D%2F'%20%7Bquote%7D%7Bdelimiter%7D%7Bjavascript%7D%2F%2F%20'%3C%7Bjavascript%7Dconfirm(1)%2F%2F'%20'%7C%7Bjavascript%7Dconfirm('xss')%2F%2F'%20'%5E%7Bjavascript%7D%5Econfirm%2F%2Fxss'%5E%2F%7Bjavascript%7D%5E%20if%20(a%20%3E%20b)%7Bxnxx_func('%3Cstrong%3E')%3B%7D%7Dalert()%3Bif(true)%7B('%3C%2Fstrong%3E')%3B%7D%20')%3B%250a%7D%250d%7D%2509alert()%3B%2F*https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io*%2Fif(true)%7B%2F%2Fhttps%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%250a('%20%3C%2FscRipT%7B%3Ffiller%7D%3E%7Bhtml%20https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%7D%20%3Ca%20href%3D%22https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%22%3EModSecurity%3C%2Fa%3E%20%3Ca%20href%3D%22https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%22%3EWordfence%3C%2Fa%3E%20%3Ca%20href%3D%22https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%22%3ECloudflare%3C%2Fa%3E%20%3Ca%20href%3D%22https%3A%2F%2Ffa4c-2600-1011-b167-9ae-808b-ab12-1ec4-bcf9.ngrok.io%22%3EAkamai%3C%2Fa%3E\",\"root_type\":\"discussion\",\"comment_id\":63450,\"comment_text\":\"%7B%22ops%22%3A%5B%7B%22insert%22%3A%22%3Cr%3E%3Ctd%3E%3Cdiv%20align%3D%5C%22center%5C%22%3E%3Ca%20href%3D%5C%22javascript%3Aconsole.log(document.cookie)%5C%22%3EHackerOne%5Cn%3C%7Btag%7D%7Bfiller%7D%7Bevent_handler%7D%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Bjavascript%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%5Cn%3CdEv%20-%20If%20fails%2C%20%3C%5Ba-zA-Z%5D%2B%20confirm(1)%5Cn%3Cd3V%20-%20If%20fails%2C%20%3C%5Ba-zA-Z0-9%5D%2Balert(1)%20%5Cn%3Cd%7C3v%20-%20If%20fails%2C%20%3C.%2Balert(1)%5Cn%3Csvg%20-%20If%20passes%2C%20no%20tag%20checking%20is%20in%20place%5Cn%3Cdev%20-%20If%20fails%2C%3C%5Ba-z%5D%2Balert(1)%5Cnx%3Cdev%20-%20If%20passes%2C%5E%3C%5Ba-z%5D%2Bconfirm(1)%5Cn%3Ctag%250dxxx%3E%20-%20If%20fails%2C%20%5B%5C%5Cs%5C%5Cn%5C%5Cr%2B%5D%2Bconfirm(1)%5Cn%3Ctag%2F~%2Fxxx%20-%20If%20fails%2C%20.%2B%5Cn%3Ctag%20xxx%20-%20If%20fails%2C%20%7Bspace%7D%5Cn%3Ctag%2509xxx%20-%20if%20fails%2C%20%5B%5C%5Cs%5D%5Cn%3Ctag%2509%2509xxx%20-%20if%20fails%2C%20%5C%5Cs%2Balert(1)%5Cn%3Ctag%2Fxxx%20-%20If%20fails%2C%20%5B%5C%5Cs%2F%5D%2Bconfirm(1)%5Cnon%5C%5Cw%2B%3Dconfirm(1)%5Cnon(load%7Cclick%7Cerror%7Cshow)%3Dconfirm(1)%5Cn%3Ctag%7Bfiller%7Donxxx%20-%20If%20fails%2C%20on%5C%5Cw%2B.%20If%20passes%2C%20on(load%7Cclick%7Cerror%7Cshow)%5Cn%3Ctag%7Bfiller%7Donclick%20-%20If%20passes%2C%20no%20event%20handler%20checking%20regular%20expression%20is%20in%20place%5Cn%3Cpayload%250aconfirm(1)%5Cn%3Cpayload%250d%20prompt(1)%5Cn%3Cpayload%2509alert(1)%5Cn%3Cpayload%3Eprompt(1)%5Cn%3Cpayload%20alert(1)%5Cn%3Cpayload%7Bspace%7Dalert(1)%5Cn%3Cpayload%2F%2Falert(1)%5Cn%3Ctag%7Bprompt(1)%7D%7Bconfirm(2)%7D%3E%5Cn%3Ca%20href%3D'%2Bconfirm(document.cookie)'%20onload%3Dconfirm(1)%3E%5Cn%3CsCriPt%7Bfiller%7DsRc%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Burl%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%5Cn%3CobJecT%7Bfiller%7Ddata%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Burl%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%5Cnhref%5B%5C%5Cs%5D%7B0%2C%7D%3D%5B%5C%5Cs%5D%7B0%2C%7Djavascript%3Aonoload%3Dconfirm(1)%5Cn%3CA%7Bfiller%7DhReF%7B%3Ffiller%7D%3D%7B%3Ffiller%7DJavaScript%3A%7Bjavascript%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%5Cn%3CA%7Bfiller%7DhReF%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Bquote%7D%7Bspecial%7D%3A%7Bjavascript%7D%7Bquote%7D%7B%3Ffiller%7D%7B%3E%2C%2F%2F%2CSpace%2CTab%2CLF%7D%5Cnj%250aAv%250dasCr%2509ipt%3Aconfirm(1)%5CnJ%250aa%250av%250aa%250as%250ac%250ar%250ai%250ap%250aT%250a%3Aprompt(1)%5CnJ%250aa%250dv%2509a%250as%250dc%2509r%250ai%250dp%2509T%250d%250a%3Aalert(1)%5Cn%26%2374%3Bavascript%26colon%3Bconfirm(1)%5Cnjav%26%23x61%3B%26%23115%3Bcript%3Aconfirm(1)%5Cn%26%2374%3Bava%250a%250d%2509script%26colon%3B%5Cn%C2%A0%3C--%24input--%3Ealert(1)%3C--%2F%24input--%3E%5Cn%3C%2Ftag%3Ealert(1)%5Cn%3C%2FtAg%2Fx%3Eprompt(1)%5Cn%3C%2Ftag%7Bspace%7D%3Eprompt(1)%5Cn%3C%2Ftag%2F%2F%3Econfirm(1)%5Cn%3C%2Ftag%250a%3Econfirm(1)%5Cn%3C%2Ftag%250d%3Ealert(1)%5Cn%3C%2Ftag%2509%3Eonerror%3Dalert(1)%5Cn%7Bclosing%20tag%7D%7Balert(1)%7D%5Cn%3Cinput%20value%3D%5C%22%24input%5C%22%20type%3D%5C%22text%5C%22%3E%5Cn%3Cscript%20src%3D%5C%22%24input%5C%22%3E%5Cn%3Cscript%20src%3D%5C%22%2Bconfirm(document.cookie)%5C%22%3E%5Cn%3Ciframe%20srcdoc%3D%5C%22%24input%5C%22%3E%5Cn%2F%2F%2Bconfirm(document.cookie)%20bypasses%20http(%3Fs)%3A%2F%2F%5Cn%2F%2F%2F%2F%2F%2F%2F%2F%2Bconfirm(document.cookie)%20bypasses%20(%3F%3Ahttp(%3Fs)%3A%3F)%3F%2F%2F%5Cn%2F%5C%5C%2F%2F%2F%5C%5C%5C%5C%2F%2Bconfirm(document.cookie)%20bypasses%20(%3F%3Ahttp(%3Fs)%3A%3F)%3F%2F%2F%2B%5Cn%3Ciframe%20srcdoc%3D%5C%22%26lt%3Bsvg%2Fonload%3Dalert()%26gt%3B%5C%22%3E%5Cn%3Cinput%20type%3D%5C%22text%5C%22%20value%3D%5C%22%5C%22%2Fonfocus%3D%5C%22alert()%24input%5C%22%3E%5Cn%7Bquote%7D%7Bfiller%7D%7Bevent_handler%7D%7B%3Ffiller%7D%3D%7B%3Ffiller%7D%7Bjavascript%7D%5Cn%7Bquote%7D%3E%7Bonclick%3Dconfirm(1)%7D%5Cnvar%20name%20%3D%20'%24input'%3B%5Cn%7Bquote%7D%7Bdelimiter%7D%7Bjavascript%7D%7Bdelimiter%7D%7Bquote%7D%5Cn'%25%7Bjavascript%7D%25'confirm(1)'%25%7B%2Fjavascript%7D%25'%5Cn'%7C%7Bjavascript%7D%7C'alert(1)'%7C%7B%2Fjavascript%7D%7C'%5Cn'%3C%7Bjavascript%7D%3C'prompt(1)'%3C%7B%2Fjavascript%7D%3C%5Cn'%3E%7Bjavascript%7D%3E'confirm(1)'%3C%2F%7Bjavascript%7D%3C%5Cn'%5E%7Bjavascript%7D%5E'alert(1)'%3C%2F%7Bjavascript%7D%3C%5Cn'*%7Bjavascript%7D*'prompt(1)'%3C%7B%2Fjavascript%7D%3C%5Cn'%2B%7Bjavascript%7D%2B'confirm(1)'%2B%2F%7Bjavascript%7D%2B'%5Cn'%2F%7Bjavascript%7D%2F'alert(1)'%2F%7Bjavascript%7D%2F'%5Cn%7Bquote%7D%7Bdelimiter%7D%7Bjavascript%7D%2F%2F%5Cn'%3C%7Bjavascript%7Dconfirm(1)%2F%2F'%5Cn'%7C%7Bjavascript%7Dconfirm('xss')%2F%2F'%5Cn'%5E%7Bjavascript%7D%5Econfirm%2F%2Fxss'%5E%2F%7Bjavascript%7D%5E%5Cnif%20(a%20%3E%20b)%7Bxnxx_func('%3Cstrong%3E')%3B%7D%7Dalert()%3Bif(true)%7B('%3C%2Fstrong%3E')%3B%7D%5Cn')%3B%250a%7D%250d%7D%2509alert()%3B%2F*%2Bconfirm(document.cookie)*%2Fif(true)%7B%2F%2F%2Bconfirm(document.cookie)%250a('%5Cn%3C%2FscRipT%7B%3Ffiller%7D%3E%7Bhtml%20%2Bconfirm(document.cookie)%7D%5Cn%3Ca%20href%3D%5C%22%2Bconfirm(document.cookie)%5C%22%3EModSecurity%3C%2Fa%3E%5Cn%3Ca%20href%3D%5C%22%2Bconfirm(document.cookie)%5C%22%3EWordfence%3C%2Fa%3E%5Cn%3Ca%20href%3D%5C%22%2Bconfirm(document.cookie)%5C%22%3ECloudflare%3C%2Fa%3E%5Cn%3Ca%20href%3D%5C%22%2Bconfirm(document.cookie)%5C%22%3EAkamai%3C%2Fa%3E%5Cn%22%7D%5D%2C%22___isMarkdown%22%3Afalse%7D\",\"chain_id\":\"ethereum\",\"author_address\":\"0xd84e5Fcd8A288AAC0Ca8fDaa949F5B34bce38960\",\"author_chain\":\"ethereum\"}",
            "category_id": "new-comment-creation",
            "view_count": 8,
            "comment_count": "2",
            "reaction_count": "0",
            "commenters": [
                {
                    "id": 118963,
                    "user_id": 121710,
                    "created_at": "2023-07-15T11:19:00.313Z",
                    "updated_at": "2023-07-15T11:19:00.315Z",
                    "profile_name": "Nicholas Harrison",
                    "email": null,
                    "website": null,
                    "bio": null,
                    "is_default": false,
                    "avatar_url": "https://lh3.googleusercontent.com/a/AAcHTtftRjY9A2G6Naypvz2ymB2ygjlXKAfDFJddmMG7vmfk=s96-c",
                    "slug": null,
                    "socials": null,
                    "background_image": null,
                    "UserId": 121710,
                    "Addresses": [
                        {
                            "id": 142533,
                            "address": "cosmos1kk26lcgeljs6f4vh6ascsafvp05ex44y8cu5zc",
                            "chain": "osmosis",
                            "role": "member",
                            "is_user_default": false,
                            "verified": "2023-07-15T11:19:00.327Z",
                            "keytype": null,
                            "last_active": "2023-07-15T11:19:00.327Z",
                            "user_id": 121710,
                            "is_councillor": false,
                            "is_validator": false,
                            "ghost_address": false,
                            "profile_id": 118963,
                            "wallet_id": "magic",
                            "UserId": 121710
                        },

```

**New:**
```
    "result": [
        {
            "thread_id": 11875,
            "last_activity": "2023-07-18T17:14:44.614Z",
            "notification_data": "{\"created_at\":\"2023-07-18T17:14:41.794Z\",\"thread_id\":11875,\"root_title\":\"6%2F23%20New%20Vitalik%20just%20dropped%20-%20Dillon%2C%20Zak%20and%20Forest\",\"root_type\":\"discussion\",\"comment_id\":63285,\"comment_text\":\"%7B%22ops%22%3A%5B%7B%22insert%22%3A%22test%20comment%206%5Cn%22%7D%5D%2C%22___isMarkdown%22%3Afalse%7D\",\"chain_id\":\"cmn-protocol\",\"author_address\":\"0x19C69A64cDA8ca5D11d1b24DD954Ec843954b598\",\"author_chain\":\"cmn-protocol\"}",
            "category_id": "new-comment-creation",
            "view_count": 46,
            "comment_count": 6,
            "commenters": [
                {
                    "id": 111299,
                    "Addresses": [
                        {
                            "id": 129605,
                            "address": "0x19C69A64cDA8ca5D11d1b24DD954Ec843954b598",
                            "chain": "cmn-protocol",
                            "profile_id": 111299
                        }
                    ]
                },
                {
                    "id": 15620,
                    "Addresses": [
                        {
                            "id": 136417,
                            "address": "0x1EFA57F404256F242175BfBd1878cAA1C36475C2",
                            "chain": "cmn-protocol",
                            "profile_id": 15620
                        }
                    ]
                },
```

rbennettcw commented:
Ok I think I understand what's going on in the PR (correct me if I'm wrong)–– the goal is to convert dynamically computed columns into persisted columns?

So with this new approach, it's updating a count column at write time via Postgres triggers, and we're storing the count in a real column, which requires no computation at read time– correct?

I think this is great in terms of a performance optimization. My main concerns are around visibility and our ability to extend functionality:
- Regarding visibility, I mean that engineers may not immediately understand how counts are updated, because that logic is written in a migration, and it wouldn't be apparent within app code.
  - Migrations don't show the current state of the DB- only the incremental changes that occurred.
  - Are there tools or anything that would allow engineers to see the latest state of all DB triggers, for debugging purposes?
- Since the trigger/computation is basically hardcoded within a migration, how easy would it be for us to change existing triggers or add new triggers if we needed to? For example, if a new reaction type was added.

Ultimately, it looks like we're trading maintainability for performance. The PR code looks good, however I think a decision needs to be made of if we want to make this tradeoff or not.

NakulManchanda commented:
@rbennettcw moving triggers to 
```
hooks: {
    afterCreate: (instance, options) => {
```

Trying out this option as well. Thanks!

jnaviask commented:
@rbennettcw's concern here, about the tradeoff between performance and maintainability, is exactly right. However, I think there's a middle ground, based on [this stackoverflow answer](https://stackoverflow.com/a/33136511): we may be able to represent the trigger as a hook in the model definition, which would expose the logic to the backend developer, perhaps at a small performance cost vs keeping the logic internal to the database. This strategy may allow us to preserve a large % of the performance gains without sacrificing maintainability, although we will need extensive docs around the hooks to communicate how/why they exist.

What do you think @NakulManchanda ?

NakulManchanda commented:
@jnaviask @rbennettcw

I've u**pdated this PR to include hooks and have also added relevant tests.**

I've introduced a new endpoint `/recomputeCounts`, which is intended solely for admin access. It might be beneficial to run this on a daily schedule to ensure column consistency in case of any divergence in count columns.

**Do we have existing admin endpoints for reference?** 
If so, I'd appreciate it if you could share an example. 
This will help me model `/recomputeCounts` using similar permissions.

rbennettcw commented:
@NakulManchanda Here's an example of the latest permissions pattern: https://github.com/hicommonwealth/commonwealth/blob/eaf3c5450271099060f7fc05cfc7c718f220ad6f/packages/commonwealth/server/controllers/server_threads_methods/archive_or_unarchive_thread.ts#L33

`validateOwner` can check if the user owns an entity or if the user is a mod/admin of a chain. If none of the conditions are true, then it throws an error.

NakulManchanda commented:
@rbennettcw -> 
is there any role with overall commonwealth admin rather than just a chain admin?

Will this check works here?
https://github.com/hicommonwealth/commonwealth/blob/b40ccd9bf67dada5d8d77109ed6d66a1b97f4de9/packages/commonwealth/server/routes/recomputeCounts.ts#L22

rbennettcw commented:
> @rbennettcw -> is there any role with overall commonwealth admin rather than just a chain admin?
> 
> Will this check works here?
> 
> https://github.com/hicommonwealth/commonwealth/blob/b40ccd9bf67dada5d8d77109ed6d66a1b97f4de9/packages/commonwealth/server/routes/recomputeCounts.ts#L22

That would be `allowGodMode: true`

rbennettcw commented:
The new hook approach looks great.

My scary question: do the hooks fire when the operation (e.g. create comment) is done via raw SQL query, or they only fire when the operation is performed via ORM?

If only triggered via ORM operations, then that's probably okay, since– from what I've seen– all/most of the CUDs are done via ORM, while queries are sometimes raw SQL.

Nonetheless, documentation should definitely note this.

NakulManchanda commented:
> My scary question: do the hooks fire when the operation (e.g. create comment) is done via raw SQL query, or they only fire when the operation is performed via ORM?

You are right I believe that too.

Here's my understanding and recommendation:
1. **Hooks & Raw SQL:** Indeed, hooks are typically not activated by operations executed via raw SQL queries. This fact emphasizes the necessity of maintaining our recompute query/endpoint.

2. **Data Capture with Raw Queries:** When employing raw queries, it's crucial to capture and handle affected `thread_id`, `comment_id`, or `notification_id` values to guarantee precise updates.

3. **Migrations & Recompute:** If we have migrations influencing data within these tables, coupling those with a subsequent recompute would ensure data consistency.

These columns essentially act as a pre-computed cache, prone to becoming stale. We have to manage this de-normalized schema ourselves.

**Given the efficiency of the recompute process, scheduling it to run, say, once a day, would serve as a safeguard.** This ensures data consistency even if we forgets recompute.

Triggers were optimal for this purpose. However, I concur that they were challenging to maintain and debug.
Moreover, they weren't sitting well with our tests as well, where we have to constantly destroy schema in un-expected ways.

NakulManchanda commented:
time to run migration on frick:

**Overall:** 26s
== 20230816015303-add-counts-thread-comment-reaction: migrated (26.826s)

**Breakdown:**
Add columns: 23.602ms
Add comment reaction count: 9.981s
Add thread reaction count: 8.651s
Add thread comment count: 6.437s
Add thread max notification id: 1.714s

rbennettcw commented:
I tested the following on Frick:
- like/unlike thread
- like/unlike comment
- create/delete comment

✅ All of them appeared to update the reaction/comment counts properly, in both the feed and thread view page.

---

However, I noticed a frontend bug:
- In the topics view, it shows 0 comments even if there are comments
- The active threads API payload correctly reports the number of comments, so it appears to be a frontend issue (cc @mzparacha)– this is only on Frick. Renders correctly on production.

<img width="1238" alt="CleanShot 2023-08-17 at 10 53 31@2x" src="https://github.com/hicommonwealth/commonwealth/assets/126922418/eec22ba5-e5f4-4f5a-aaaa-7c282980c35e">

NakulManchanda commented:
> In the topics view, it shows 0 comments even if there are comments

Let me check

NakulManchanda commented:
Hi @rbennettcw 
- did fix those counts.
- but as fast follow or with this PR we can refactor frontend to use `comment_count`, eliminating the need for this extra map.

```javascript
const allRecentTopicThreads = allRecentTopicThreadsRaw.map((thread) => {
    const t = thread.toJSON();
    t.numberOfComments = t.comment_count || 0;
    return t;
  });
```

jnaviask commented on packages/commonwealth/server/migrations/20230816015303-add-counts-thread-comment-reaction.js:
```diff
@@ -0,0 +1,115 @@
+      );
+      console.timeEnd('Add comment reaction count');
+
+      // TODO: add comments reaction
```
Open TODO

NakulManchanda replied:
removed

jnaviask commented on packages/commonwealth/server/routes/recomputeCounts.ts:
Under what circumstances would we call this route?

NakulManchanda replied:
i think we should call it on schedule, once a day, just so numbers don't get out of sync. It's a safe guard mechanism.
Here is complete discussion:
https://github.com/hicommonwealth/commonwealth/pull/4200#issuecomment-1681021679

Lets figure out best way to schedule this one!

jnaviask replied:
@NakulManchanda why would the data become stale over time? Under what circumstances would an increment get missed and the count become inaccurate?

NakulManchanda replied:
using hook is not full proof, 
- its possible data get changed using raw queries outside of app in migration scripts, direct raw queries etc, some un-forseen failure. 
- Just a failsafe for those scenarios. Also, to avoid race condition which can occure.

NakulManchanda replied:
Basically, it can be used to recompute counts on demand. Even if we don't want to schedule a job, we can just keep this endpoint around in case we find issue(s) with counts and would like to refresh them.

NakulManchanda replied:
- removed recompute count endpoint

NakulManchanda replied:
added back

jnaviask reviewed (commented):
I'm generally happy with the structure of the improvements. Left a few Qs and awaiting @rbennettcw signoff + a round of QA, then should be good to go.

NakulManchanda commented:
>  a round of QA, then should be good to go.

Deployed on Frick for click around testing
https://commonwealth-frick.herokuapp.com/

NakulManchanda commented:
> what's the use of max notification ID?

most recent - activity on new-thread & new-comment notification.
It speed up feed queries
`viewUserActivity` & `viewGlobalActivity`

NakulManchanda commented:
@rbennettcw if ok, can you mark it approved?

rbennettcw reviewed (approved):
Just want to note:

The hooks solution works great for logic that uses ORM methods. However, the hooks are not trigged by raw SQL queries (anything that mutates the DB aside from the ORM).

Example:

This WILL update reaction counts:
```ts
const result = await models.Reaction.findOrCreate(…)
```

This will NOT update reaction counts:
```ts
const result = await sequelize.query('INSERT INTO Reactions…')
```

This is why the recompute endpoint wasn't a terrible idea as a safeguard.

---

Overall, the current solution is fine because all relevant DB queries (e.g. create comment, create reaction) use the ORM.

I don't foresee the caveat being a real issue anytime soon.

jnaviask commented:
@NakulManchanda  As discussed in parking lot, to merge:
- Re-add the recompute endpoint + job at 3 AM
- Add new ticket for reporting metrics on change when recomputing
- Recompute job should use secret rather than JWT approach

NakulManchanda commented:
@jnaviask 

> Re-add the recompute endpoint
- done

> + job at 3 AM  
> Recompute job should use secret rather than JWT approach

**As a fast follow** - i can add new strategy to passport - `passport-http-bearer`
- rename endpoint to /api/job/recomputeCounts
- we can add any future job to `/api/job` prefix and call them from Heroku Schechduler like this
```
curl -H "Authorization: Bearer $SECRET_TOKEN" -X POST https://commonwealth.im/api/job/recomputeCounts
```

> Add new ticket for reporting metrics on change when recomputing

**As a fast follow** i can add following new metrics 
- cw.hook.commentCount.inc
- cw.hook.commentCount.dec
- cw.hook.reactionCount.inc
- cw.hook.reactionCount.dec
- cw.hook.maxNotif

or something simillar we can discuss under new ticket

NakulManchanda commented:
Lets do two follow up items
- one using secret instead of jwt - lets do this one properly too in separate PR
- one adding datadog metrics to keep track of activity from hooks

NakulManchanda commented:
using secret instead of jwt:
https://github.com/hicommonwealth/commonwealth/pull/4873

jnaviask commented:
@NakulManchanda final steps for this PR and then good to go:
- ensure all CI flows pass (saw that unit / integration tests failing)
- please write up on the main PR body: what specific steps required to initialize the recompute job? what heroku config (env vars) / scripts needed?
- feel free to open another ticket to add metrics as a fast follow
- need a more detailed test plan for QA to ensure that code works as expected

NakulManchanda commented:
> ensure all CI flows pass (saw that unit / integration tests failing)

Can we add new environment vars to CW_SCHEDULER_SECRET to github secrets?
This will make unit / integration tests pass

> please write up on the main PR body: what specific steps required to initialize the recompute job? what heroku config (env vars) / scripts needed?

Updating

> feel free to open another ticket to add metrics as a fast follow

Ticket added yesterday - #4867 
PR opened today - #4879

> need a more detailed test plan for QA to ensure that code works as expected

- Can deploy to Frick/Frack - for QA testing
- Running load test to see if its holding well against Frick

jnaviask commented:
> Can we add new environment vars to CW_SCHEDULER_SECRET to github secrets?
> This will make unit / integration tests pass

Done.

> Can deploy to Frick/Frack - for QA testing

QA testing happens on the QA server. Your code will automatically be deployed there once merged. QA testing is performed by the product team -- not by engineering, so we need to provide for them a plan on how to verify the changes are successful.

NakulManchanda commented:
> * Re-add the recompute endpoint + job at 3 AM
it was added based on our conversation from Tuesday
And merged after review from both you and Ryan
https://github.com/hicommonwealth/commonwealth/pull/4873

> Last thing is to add script in package.json in server/scripts folder instead of recompute job route, then good to go.

Its ok we can get rid of it
- should we delete all the related code to handle secrets - or just `recompute` route?
-  if we envision using secrets for anything in near future - we can keep code else we can revert whole PR 4873 and more

jnaviask commented:
I can see no specific reason to keep any secret related code around.

NakulManchanda commented:
> I can see no specific reason to keep any secret related code around.

- changed to script, code is deleted
- instructions - to run from root folder
```
yarn --cwd packages/commonwealth recompute-counts
```
We need to find a best way to run it on Heroku scheduler.

NakulManchanda commented:
tried jobs on Frick, they are working:

<img width="1328" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/f1332940-7b55-493c-b75a-c566f6557498">

Job Logs:
<img width="1271" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/4791635/3b215d12-4263-436d-9418-0bcf5496da5d">

NakulManchanda commented:
Added two sample jobs and general scripts
- recompute-count-job.ts
- recompute-count-query.sql 

Added - so people can refer general structure if they like to create raw sql

Both works fine. 
Lets us know which one we would like to keep, we can delete the other one.

timolegros commented:
> > Last thing is to add script in package.json in server/scripts folder instead of recompute job route, then good to go.
> 
> Added new job
> 
> ```
> yarn --cwd packages/commonwealth recompute-counts
> ```
> 
> Tested on heroku on frick, it works fine.
> 
> `Database cleaner` process can be migrated to new process in similar way.

I'd rather not alter the database cleaner to use the Heroku scheduler. Using Heroku Scheduler introduces database side-effects for which there is no code (as long as we don't use IAC). IMO a Heroku scheduled job is less visible than database-level triggers which at least there is a migration for (and it shows up in database explorers) and less visible than an in-memory scheduled job for which there is code.

If we still go with Heroku Scheduler then it needs to be rigorously documented to make it excessively clear.

Also, I don't see any tests for the recompute job so it could become the case that the schema changes (while recompute job raw queries are not updated) so the scheduled job begins to fail in production which is not something that will be caught in QA unless the job schedule is updated to run precisely during QA.

jnaviask commented:
Determination here is that:
- Where we land on the repeating job wont fully block deployment, as job is not strictly necessary (but we have no estimation of how much we expect the counts to diverge from the real numbers?).
- We will want to ship a repeating job as a fast follow. @timolegros and @NakulManchanda should discuss on a separate ticket how to approach a repeating job, because their opinions differ. Would like to at least see an approach decided or some conversation happening before we merge, because we will want to ship some sort of job within a day or two of shipping the rest of the code changes.

NakulManchanda commented:
I don't see anything above is a blocker for current PR.
**How to schedule job is a separate discussion, and we can mature on our current solution as separate task.**

> Heroku scheduled job is less visible than database-level triggers  

Heroku scheduled jobs and database-level triggers serve different purposes; they are not directly comparable

> Heroku Scheduler introduces database side-effects for which there is no code (as long as we don't use IAC). 

What is a database side-effect?
It is true for any piece of code modifying data.

> any tests for the recompute job specifically 

It's possible to implement tests for jobs like the recompute job. Incorporating such tests can be considered for future tasks.

>  the schema changes (while recompute job raw queries are not updated)  

- Schema changes impact various parts of the codebase, not just specific job.
- All hooks introduced in PR are rlying on same schema as well
- scope of the job is limited to the new columns we introduced and is already reviewed.

> so the scheduled job begins to fail in production

- This job is exact replica of migration script, running migration script which is already reviewed on schedule
- This job in particular is a fail safe, if it fails it is not a P1 issue
- This job in particular, it's to avoid race conditions.
- This is not a must have job, good to have job - so its ok to ship without this

In general, scheduling externally is better than using javascript `settimeout`

 `settimeout`
- it can be unreliable, especially during server restarts. 
- Coordinating tasks across multiple servers using locks in Redis can be complex.
- In-house solutions for in-memory scheduling lack maturity in terms of how to time jobs across multiple server and lacks visibility.

`scheduling externally` 
- Operates on a separate dyno, ensuring the main server's compute resources are not utilized.
- Offers superior scheduling capabilities.
- Provides a dashboard to monitor all scheduled jobs and their last execution times.

## Addressing limitation of Heroku Scheduler
To address the limitations of Heroku Scheduler, alternatives like crontogo can be explored. 
https://devcenter.heroku.com/articles/crontogo

Some of its features include:
- An extensive job history dashboard.
- A grid to monitor job successes and failures, enhancing visibility.
- Support for shorter time intervals (1 minute compared to Heroku's 10 minutes).
- Easy migration of Heroku Scheduler job (we have 1 job, so not a big feature for us)
- More granular control over scheduling using cron syntax.
  
Let's discuss any potential issues and address any doubts.
Current solution using Heroku scheduler, it appears to be a reliable and user-friendly solution for scheduling tasks, and could be used repeatedly without any concern.
And we can migrate any in-memory scheduled job we have, like refreshing cache, archiving notifications etc.

**Testing is a separate concern and is not directly related to the job's execution environment.** 
Enhancing testing is always beneficial, and feedback is well taken.

jnaviask commented:
cc @NakulManchanda to fix merge conflicts
